### PR TITLE
Ignore const fn in `or_fun_call` lint

### LINKED ISSUE #5658: or_fun_call triggers on const fn
```rust
const fn f(v: i32) -> Option<i32> {
    Some(v)
}

fn main() {
    None.or(f(1));  // try this: `or_else(|| f(1))`
}
```

### PR DESCRIPTION
**Pull request by lzutao**
cc  https://github.com/rust-lang/rust-clippy/issues/1609#issuecomment-562929047
Closes #5658

changelog: none

### CODE REVIEW

File: clippy_lints/src/utils/mod.rs [hunk 1]
```diff
@@ -911,6 +911,7 @@ pub fn is_ctor_or_promotable_const_function(cx: &LateContext<'_, '_>, expr: &Exp
         if let ExprKind::Path(ref qp) = fun.kind {
             let res = cx.tables.qpath_res(qp, fun.hir_id);
             return match res {
+                def::Res::Def(DefKind::Fn, def_id) => rustc_mir::const_eval::is_const_fn(cx.tcx, def_id),
```
Comment (lzutao) [inline clippy_lints/src/utils/mod.rs:914]: There are another way: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/struct.FnHeader.html#structfield.constness

Comment (oli-obk) [reply clippy_lints/src/utils/mod.rs:914]: No, this is definitely the right way. The constness field by itself is set, even if the function's constness is unstable.

**Comment by oli-obk** [comment]
I don't think the constness of the function by itself will give us what we want. If the function is a const fn, but has arguments, these arguments may not be constants. An example:

```rust
const fn compute_nth_prime(n: usize) -> Option<usize> {
    // super heavy logic here
}
fn foo(n: usize) -> Option<usize> {
	// Whether the function above is `const fn` or not has no effect on the cost
    // of the work done in the `or`.
	None.or(compute_nth_prime(n))
}
```

Thus, we'll end up actually running the function at runtime, with all the runtime effort that comes with it. If you really want to do this, you'll need to analyze the entire expression and check if it is entirely const.

I don't know of a reliable way to do this except to do it on MIR, which would be a major undertaking.

**Comment by matthiaskrgr** [comment]
This is a user-facing change, the "changelog:" line should say which false positive this fixes.

**Comment by lzutao** [comment]
cc @ecstatic-morse do you know how to detect if a const fn could be evaluated at compile-time.

**Comment by oli-obk** [comment]
This is not about whether it *could* be evaluated at compile-time. A `const fn` by definition can be evaluated at compile-time. It's whether your expression can be evaluated at compile time. I think the cleanest way would be to implement `const fn` const prop in rustc and make this lint a MIR lint that just checks whether the argument to `or` is a const (or one of the whitelisted kinds of functions here).

**Comment by lzutao** [comment]
I don't think I have enough skill to make that change.
Closing this PR in the mean time.

**Pull request closed without merging**